Gamepad

UvuvDev opened the pull request:
#### Summary

A controller wrapper around the PROS Controller that adds extra functionality, such as shift keys, toggles, controller map modes, and auto-running functions.

#### Motivation

Add extra functionality to both the controller and LemLib.

#### Test Plan

Unit tests, and running it on the bot


<!-- DO NOT REMOVE!! -->
<!-- bot: nightly-link -->
<!-- commit-sha: 1a61cf5e6bbc17f48ff00d4dab143de5a76044af -->
## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/7591652397)
- via manual download: [LemLib@0.5.0-rc2+1a61cf.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1182552472.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc2+1a61cf.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1182552472.zip;
pros c fetch LemLib@0.5.0-rc2+1a61cf.zip;
pros c apply LemLib@0.5.0-rc2+1a61cf;
rm LemLib@0.5.0-rc2+1a61cf.zip;
```

UvuvDev committed:
Acorn Tracking Implemented

Acorn Tracking is ported from uvuvlib. Derived from movement, it gets you the coordinates of the leftmost acorn seen.

liam-teale committed:
Merge branch 'refactor' into refactor

liam-teale committed:
Merge branch 'refactor' into refactor

UvuvDev committed:
New Controller Wrapper, LEMController

LEMController has the functionality of a PROS controller, but also supports shift keys, toggles, new button presses, and automatically running functions on button presses

UvuvDev committed:
Merge branch 'refactor' of https://github.com/DancingRicardo/LemLib into refactor

UvuvDev committed:
Did some edits to controller, Added Reader Auton

Reader Auton, can use Lem Drivetrains or user's custom functions to move the drive based off previous inputs. Custom functions now require an int parameter, if user doesn't want to use it it can just be left blank. just adds more options.

UvuvDev committed:
Unit Tests for controller

added unit tests for controller functions.

UvuvDev committed:
AccuracySIM psuedocode

currently non functional, written in psuedocode. Class is in place, will work on inheriting from lemlib movement class soon.  Have to port the Poller from UvuvLib as well.

UvuvDev committed:
Bug Fixes, LEMControllerValues Struct

Fixed bugs, mainly bugs stemming from not saving button states to memory. This meant that functions could keep track of one button instead of 12 like its supposed to.

This was solved by introducing a hashmap of button types and booleans, which kept track of each individual buttons state. LEMControllerValues was uncommented and reformatted to grab the keys based off a button parameter, meaning the code will now be much cleaner.

UvuvDev committed:
Bug Fix, Unit test fixes

Previously, controller was unable to run functions when buttons were false, which would mean you lose half the functionality. Fixed by making every function pointer in the code a pair now, first being for when the function is false, the second being for when it is true.

This bug fix meant some of the unit tests needed to be redone, which they are now.

Raptorly1 committed:
clang-format

UvuvDev committed:
Merge pull request #1 from Raptorly1/AcornTracker-clang-format

UvuvDev committed:
Started PID tuner

Started work on the PID Tuner class. Has a constructor, FF and kP functionality is done for flywheels.

UvuvDev committed:
Gamepad

UvuvDev committed:
Took out other PR's

UvuvDev committed:
Forgot a few header files, oops

UvuvDev committed:
Fixing Merge Conflicts

UvuvDev committed:
Fixed compile error

fixed error with double definiton

UvuvDev committed:
Merge branch 'refactor' into master

UvuvDev committed:
format

UvuvDev committed:
Merge branch 'master' of https://github.com/DancingRicardo/LemLib

UvuvDev committed:
format again

liam-teale commented on include/lemlib/devices/gamepad.hpp:
```diff
@@ -0,0 +1,283 @@
+
+namespace lemlib {
+
+struct LEMButtonMapping {
```
Not necessary to preface names with `LEM`

liam-teale commented on include/lemlib/devices/gamepad.hpp:
```diff
@@ -0,0 +1,283 @@
+#include "pros/device.hpp"
+#include "unordered_map"
```
```suggestion
```
you have it included below

liam-teale commented on src/lemlib/movements/pursuit.cpp:
why has this been renamed?

liam-teale commented on src/lemlib/devices/gamepad.cpp:
```diff
@@ -0,0 +1,185 @@
+}
+
+void Gamepad::autoButtonFunctions() {
+    // for (int i = 0; i < buttonStates.size(); i++) {
+    //     buttonStates.at(buttonsToFunctions.at(i)->getButton()) =
+    //     prosController->get_digital(buttonsToFunctions.at(i)->getButton());
+    // }
```
you probably want to remove this

liam-teale commented on include/lemlib/devices/gamepad.hpp:
```diff
@@ -0,0 +1,283 @@
+};
+
+class Gamepad {
+    private:
```
```suggestion
```

UvuvDev committed:
Made Requested Changes, fixed Unit Tests

Unit tests hadn't been updated after renaming everything, so I edited them. Requested changes are just the one listed by Liam in the PR

UvuvDev committed:
Move all function definitions out of headers

All the structs had function definitions inside header files instead of inside source. Moved them over.

UvuvDev committed:
Switched to using this->

switched out original syntax (ex. mode = modeParam; ) to the preferred standard (ex. this->mode = mode;)

UvuvDev committed:
Switched all parameter std::strings to const std::string&

UvuvDev committed:
Raw Pointers removed, replaced

Removed all raw pointers and replaced them with shared and unique pointers

UvuvDev committed:
Changed every function pointer to std::function

UvuvDev committed:
Switched many for loops to for each

Made a lot of for loops into for each loops for code readability.
